src/controller/cpp/Robot.cpp

```diff
@@ -449,15 +449,14 @@ Device *Robot::getOrCreateDevice(int tag) {
   if (tag == 0)
     return NULL;
 
-  int size = (int)deviceList.size();
-  if (size > 0) {
-    if (size <= tag)
-      return NULL;
+  const int count = wb_robot_get_number_of_devices();
+  const int size = (int)deviceList.size();
+  if (size == count && size > 0 && tag < size)
```

Comment:
If we remove a device and add a new one during the same step, the count will remain the same, however, this lists may differ, isn't it? So, we should also re-initialize the list. It seems to me this test may not be sufficient to determine if the device list changed.

Reply:
Yes.
We need to add an additional function in the C API that returns if the device list has been changed or not.

Reply:
I checked again, and with the current implementation it is enough to check if the count changed because devices are never removed from the C controller list so the size of the devices can only increase.

The return value of `robot_get_number_of_devices` is always the maximum tag value + 1.
Even if you remove a device it the count won't change and given that tags are not reused, when you add a new device the count will for sure increase.

Reply:
OK, that sounds good.

Reply:
I would be nice nevertheless to write a comment about it in the code to make it clear.